pkg/pool-utils/contracts/lib/VaultReentrancyLib.sol

```diff
@@ -33,8 +34,24 @@ library VaultReentrancyLib {
      * here (https://forum.balancer.fi/t/reentrancy-vulnerability-scope-expanded/4345), those functions are unsafe,
      * and subject to manipulation that may result in loss of funds.
      */
-    function ensureNotInVaultContext(IVault vault) internal {
-        IVault.UserBalanceOp[] memory noop = new IVault.UserBalanceOp[](0);
-        vault.manageUserBalance(noop);
+    function ensureNotInVaultContext(IVault vault) internal view {
+        // Perform the following operation to trigger the Vault's reentrancy guard.
+        // Use a static call so that it can be a view function (even though the
+        // function is non-view).
+        //
+        // IVault.UserBalanceOp[] memory noop = new IVault.UserBalanceOp[](0);
+        // _vault.manageUserBalance(noop);
+
+        // solhint-disable-next-line var-name-mixedcase
+        bytes32 REENTRANCY_ERROR_HASH = keccak256(abi.encodeWithSignature("Error(string)", "BAL#400"));
+
+        // read-only re-entrancy protection - we are making a static call on a non-view function. It will either
+        // "succeed" (return with no transaction), or revert. if it reverts, we need to make sure it didn't fail
+        // due to a re-entrancy attack.
+        (, bytes memory revertData) = address(vault).staticcall(
+            abi.encodeWithSelector(vault.manageUserBalance.selector, new address[](0))
+        );
+
+        _require(keccak256(revertData) != REENTRANCY_ERROR_HASH, Errors.REENTRANCY);
```

Comment:
Bytecode size has been a concern, and this could likely have its bytecode size reduced greatly by using assembly, which would let us hardcode the selector, expected revert reason, calldata, etc. It might be a good idea to strengthen the tests in that case.

Alternatively we could move this to a separate contract and call there. Or we could also not care about this.

Reply:
This code is dangerous because the  `static_call` might revert for other reasons before hitting reentrancy error. We should make sure it never happens. 
The previous version is different because it requires a successful `manageUserBalance` call which is different compare to only failing with the reentrancy error. In fact would `manageUserBalance` call succeed if it doesn't change the state?